net/interfaces/interfaces_windows.go

```diff
@@ -7,76 +7,91 @@ package interfaces
 	"tailscale.com/tsconst"
-	"tailscale.com/util/lineread"
 )
 
 func init() {
 	likelyHomeRouterIP = likelyHomeRouterIPWindows
 	getPAC = getPACWindows
 }
 
-/*
-Parse out 10.0.0.1 from:
-
-Z:\>route print -4
-===========================================================================
-Interface List
- 15...aa 15 48 ff 1c 72 ......Red Hat VirtIO Ethernet Adapter
-  5...........................Tailscale Tunnel
-  1...........................Software Loopback Interface 1
-===========================================================================
-
-IPv4 Route Table
-===========================================================================
-Active Routes:
-Network Destination        Netmask          Gateway       Interface  Metric
-          0.0.0.0          0.0.0.0         10.0.0.1       10.0.28.63      5
-         10.0.0.0      255.255.0.0         On-link        10.0.28.63    261
-       10.0.28.63  255.255.255.255         On-link        10.0.28.63    261
-        10.0.42.0    255.255.255.0   100.103.42.106   100.103.42.106      5
-     10.0.255.255  255.255.255.255         On-link        10.0.28.63    261
-   34.193.248.174  255.255.255.255   100.103.42.106   100.103.42.106      5
-
-*/
 func likelyHomeRouterIPWindows() (ret netaddr.IP, ok bool) {
-	cmd := exec.Command("route", "print", "-4")
-	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
-	stdout, err := cmd.StdoutPipe()
+	rs, err := winipcfg.GetIPForwardTable2(windows.AF_INET)
 	if err != nil {
+		// TODO: It would be nice to log this...
 		return
 	}
-	if err := cmd.Start(); err != nil {
-		return
+
+	// Effective metric is sum of interface metric and route metric offset
+	setEffectiveMetric := func(r *winipcfg.MibIPforwardRow2) bool {
+		iface, err := r.InterfaceLUID.IPInterface(windows.AF_INET)
+		if err != nil {
+			// TODO: It would be nice to log this...
+			return false
+		}
+		r.Metric += iface.Metric
+		return true
 	}
-	defer cmd.Wait()
 
-	var f []mem.RO
-	lineread.Reader(stdout, func(lineb []byte) error {
-		line := mem.B(lineb)
-		if !mem.Contains(line, mem.S("0.0.0.0")) {
-			return nil
+	unspec := net.IPv4(0, 0, 0, 0)
+	var route *winipcfg.MibIPforwardRow2
+	metricUpdated := false
+
+	for _, r := range rs {
+		if r.Loopback || r.DestinationPrefix.PrefixLength != 0 || !r.DestinationPrefix.Prefix.IP().Equal(unspec) {
+			// Not a default route, so skip
+			continue
+		}
+
+		ip, ok := netaddr.FromStdIP(r.NextHop.IP())
+		if !ok || ip.IsZero() {
+			// Not a valid gateway, so skip (won't happen though)
+			continue
+		}
+
+		if route == nil {
+			route = &r
+			ret = ip
+			continue
 		}
-		f = mem.AppendFields(f[:0], line)
-		if len(f) < 3 || !f[0].EqualString("0.0.0.0") || !f[1].EqualString("0.0.0.0") {
-			return nil
+
+		// We can get here only if there are multiple default gateways defined (rare case),
+		// in which case we need to calculate the effective metric.
+		// First update the previous route metric if not already updated:
+		if !metricUpdated {
```

Comment:
can we do without this variable if `setEffectiveMetric` returns early if its Metric field is already non-zero?

Reply:
Not sure that my explanation was clear, so I'll try to clarify: I avoided updating `route.Metric` as soon as I get the first default route because most often it'll be the only default route, so I won't need `route.Metric` at all. But then if I find another default route I need to calculate it for both - current and previous.

Reply:
Well, in the latest commit I've removed `metricUpdated` variable by replacing it with `ifaceMetricCache == nil` check. `ifaceMetricCache` is a new thing that I've introduced in the latest commit...